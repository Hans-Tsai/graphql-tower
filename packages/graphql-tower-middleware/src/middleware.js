import _ from 'lodash';
import { UnauthorizedError } from 'graphql-tower-errors';

export function authentication(payload, args, context) {
  if (!_.has(context, 'user.id')) {
    throw new UnauthorizedError('authorization header is required');
  }
}

export { authentication as default };
