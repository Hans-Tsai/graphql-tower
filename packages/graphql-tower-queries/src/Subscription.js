import GraphQLField from './GraphQLField';

export default class Subscription extends GraphQLField {
  constructor() {
    super();

    Object.defineProperty(this, 'resolve', {
      enumerable: true,
      set: (resolve) => { this._.resolve = resolve; },
      get: () => this.constructor.resolve.bind(this),
    });

    Object.defineProperty(this, 'subscribe', {
      enumerable: true,
      set: (subscribe) => { this._.subscribe = subscribe; },
      get: () => async (...args) => {
        await this.constructor.middleware.apply(this, args);
        return this._.subscribe.apply(this, args);
      },
    });
  }
}
