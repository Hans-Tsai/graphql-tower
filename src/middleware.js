import _ from 'lodash';
import { UnauthorizedError } from './error';

export function authentication(payload, args, context) {
  if (!_.has(context, 'user.id')) {
    throw new UnauthorizedError('authorization header is required');
  }
}

export { authentication as default };
