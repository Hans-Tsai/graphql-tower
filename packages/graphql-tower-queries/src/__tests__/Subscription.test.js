import gql from 'graphql-tag';
import { execute, subscribe, GraphQLSchema, GraphQLObjectType, GraphQLInt } from 'graphql';
import { createServer } from 'http';
import ws from 'ws';
import { PubSub } from 'graphql-subscriptions';
import { SubscriptionServer, SubscriptionClient } from 'subscriptions-transport-ws';
import { authentication } from '../middleware';
import Subscription from '../Subscription';

const pubsub = new PubSub();

const subscribeSpy = jest.fn(() => pubsub.asyncIterator('messageAdd'));
const resolveSpy = jest.fn(payload => payload * 2);

const OnMessageAdd = class extends Subscription {
  type = GraphQLInt;
  middleware = [authentication];
  resolve = resolveSpy;
  subscribe = subscribeSpy;
};

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({ name: 'Query', fields: { onMessageAdd: new OnMessageAdd() } }),
  subscription: new GraphQLObjectType({ name: 'Subscription', fields: { onMessageAdd: new OnMessageAdd() } }),
});

const options = {
  schema,
  subscribe,
  execute,
  onConnect: params => (params.token ? { user: { id: 10 } } : {}),
};

function notFoundRequestListener(request, response) {
  response.writeHead(404);
  response.end();
}

const httpServer = createServer(notFoundRequestListener);
httpServer.listen();
new SubscriptionServer(options, { server: httpServer, path: '/' }); // eslint-disable-line
const { port } = httpServer.address();

describe('Subscription', () => {
  it('snapshot', async () => {
    const subscription = new Subscription();
    expect(subscription).toMatchSnapshot();
  });

  it('subscribe successfully', async (done) => {
    const results = await subscribe(schema, gql`subscription { onMessageAdd }`, {}, { user: { id: 10 } });
    results.next().then(({ value: { data } }) => {
      expect(data).toEqual({ onMessageAdd: 20 });
      done();
    });

    pubsub.publish('messageAdd', 10);
  });

  it('subscribe when authorization', async (done) => {
    const client = new SubscriptionClient(`ws://localhost:${port}/`, { connectionParams: { token: 'XYZ' } }, ws);
    client.request({ query: 'subscription { onMessageAdd }' }).subscribe({
      next: ({ data }) => {
        expect(data).toEqual({ onMessageAdd: 20 });
        expect(subscribeSpy)
          .toHaveBeenLastCalledWith(undefined, {}, { user: { id: 10 } }, expect.anything());
        expect(resolveSpy)
          .toHaveBeenLastCalledWith(10, {}, { user: { id: 10 } }, expect.anything());
        client.close();
        done();
      },
    });

    setTimeout(() => pubsub.publish('messageAdd', 10), 100);
  });

  it('subscribe when is guest', async (done) => {
    const client = new SubscriptionClient(`ws://localhost:${port}/`, {}, ws);
    client.request({ query: 'subscription { onMessageAdd }' }).subscribe({});
    setTimeout(() => {
      expect(subscribeSpy).toHaveBeenCalledTimes(1);
      expect(resolveSpy).toHaveBeenCalledTimes(0);
      client.close();
      done();
    }, 200);

    setTimeout(() => pubsub.publish('messageAdd', 10), 100);
  });
});
