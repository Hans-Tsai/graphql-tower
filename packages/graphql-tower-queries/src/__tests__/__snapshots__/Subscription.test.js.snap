// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscription snapshot 1`] = `
Subscription {
  "_": Object {
    "afterware": Array [],
    "middleware": Array [],
    "name": undefined,
    "resolve": [Function],
    "subscribe": [Function],
  },
  "args": Object {},
  "resolve": [Function],
  "subscribe": [Function],
}
`;
